src/System.Private.CoreLib/shared/System/Activator.RuntimeType.cs

```diff
@@ -133,20 +133,106 @@ public static partial class Activator
+                    }
+                }
+
+                return instance;
+            }
+            else
+            {
+                if (entry.ConstructClass == null)
+                    throw new MissingMethodException(SR.Format(SR.Arg_NoDefCTor, entry.RuntimeType));
+
+                T instance = (T)RuntimeTypeHandle.Allocate(entry.RuntimeType);
+
+                try
+                {
+                    entry.ConstructClass(instance);
+                }
+                catch (Exception exception)
+                {
+                    throw new TargetInvocationException(exception);
+                }
+
+                return instance;
+            }
+        }
+
+        private delegate void ClassConstructor<T>(T instance);
+        private delegate void StructConstructor<T>(ref T instance);
+
+        private static T CreateDelegate<T>(MethodBase methodInfo)
+            where T : Delegate
+        {
+            return (T)typeof(T)
+                .GetConstructor(new[] { typeof(object), typeof(IntPtr) })!
+                .Invoke(new object?[] { null, methodInfo.MethodHandle.GetFunctionPointer() });
+        }
+
+        private sealed class ActivatorCache<T>
+        {
+            internal static readonly ActivatorCache<T> Entry = new ActivatorCache<T>();
+
+            internal readonly RuntimeType RuntimeType;
+
+            internal readonly bool IsValueType;
+            internal readonly bool IsNotActivatable;
 
-            // This is a workaround to maintain compatibility with V2. Without this we would throw a NotSupportedException for void[].
-            // Array, Ref, and Pointer types don't have default constructors.
-            if (rt.HasElementType)
-                throw new MissingMethodException(SR.Format(SR.Arg_NoDefCTor, rt));
+            internal readonly ClassConstructor<T>? ConstructClass;
+            internal readonly StructConstructor<T>? ConstructStruct;
+
+            private ActivatorCache()
+            {
+                RuntimeType = (RuntimeType)typeof(T);
+                IsValueType = RuntimeType.IsValueType;
 
-            // Skip the CreateInstanceCheckThis call to avoid perf cost and to maintain compatibility with V2 (throwing the same exceptions).
-            return (T)rt.CreateInstanceDefaultCtor(publicOnly: true, skipCheckThis: true, fillCache: true, wrapExceptions: true);
+                // The HasElementType check is a workaround to maintain compatibility with V2.
+                // Without this we would throw a NotSupportedException for void[].
+                // Array, Ref, and Pointer types don't have default constructors.
+                IsNotActivatable = RuntimeType.IsAbstract || RuntimeType.HasElementType;
+
+                if (IsNotActivatable)
+                    return;
+
+                ConstructorInfo? runtimeCtorInfo = RuntimeType.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, Type.EmptyTypes, null);
```

Comment:
Would it be better to move this inside the two conditions below? That would prevent the `BindingFlags.NonPublic` and associated `runtimeCtorInfo.IsPublic` from being passed to `GetConstructor` for reference types.

Reply:
The reason why `BindingFlags.NonPublic` is used here is that `struct`s could have parameterless constructors. Therefore, we should allow struct activation only in case when it has no parameterless constructor or have it and it's public. If I get only a public constructor, then I won't know if there is a private one.

Take a look on dotnet/corefx#38032 which adds tests for these scenarios.

Reply:
Ah, sorry, I was tired and missread the comment.

```csharp
if (IsValueType)
{
    ConstructorInfo? runtimeCtorInfo = RuntimeType.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, Type.EmptyTypes, null);

    if (runtimeCtorInfo != null && runtimeCtorInfo.IsPublic)
        ConstructStruct = CreateDelegate<StructConstructor<T>>(runtimeCtorInfo);
    else
        IsNotActivatable = runtimeCtorInfo != null;
}
else
{
    ConstructorInfo? runtimeCtorInfo = RuntimeType.GetConstructor(BindingFlags.Instance | BindingFlags.Public, null, Type.EmptyTypes, null);

    if (runtimeCtorInfo != null)
        ConstructClass = CreateDelegate<ClassConstructor<T>>(runtimeCtorInfo);
    else
        IsNotActivatable = true;
}
```

Not sure that duplicated call of `GetConstructor` makes code better. Maybe the following refactoring is better? What do you think? It has only visibility check.

```csharp
ConstructorInfo? runtimeCtorInfo = RuntimeType.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, Type.EmptyTypes, null);

if (runtimeCtorInfo != null && runtimeCtorInfo.IsPublic)
{
    if (IsValueType)
        ConstructStruct = CreateDelegate<StructConstructor<T>>(runtimeCtorInfo);
    else
        ConstructClass = CreateDelegate<ClassConstructor<T>>(runtimeCtorInfo);
}
else
{
    if (IsValueType)
        IsNotActivatable = runtimeCtorInfo != null;
    else
        IsNotActivatable = true;

    // which can be simplified to
    IsNotActivatable = !(IsValueType && runtimeCtorInfo == null);
}
```